File: src/main/scala/com/bdmendes/smockito/internal/MockitoAgent.scala
package com.bdmendes.smockito.internal

import java.lang.instrument.Instrumentation
import scala.util.Failure
import scala.util.Success
import scala.util.Try

object MockitoAgent:

  def method =
    // Unfortunately, we cannot be sure that Mockito exists in the class path in
    // some complicated build configurations. It's best not to load this agent
    // in that case, but nevertheless we want to be flexible.
    Try {
      val mockito = Class.forName("org.mockito.internal.PremainAttach")
      mockito.getMethod("premain", classOf[String], classOf[Instrumentation])
    }

  def premain(args: String, instrumentation: Instrumentation): Unit =
    method match
      case Failure(e) =>
        println(s"Smockito could not start Mockito agent: ${e.getMessage}")
      case Success(m) =>
        // Mockito already checks for repeated calls. We can invoke this blindly.
        m.invoke(null, args, instrumentation)

File: src/main/java/com/bdmendes/smockito/internal/MockitoAgent.java
package com.bdmendes.smockito.internal;

import java.lang.instrument.Instrumentation;
import java.lang.reflect.Method;

public class MockitoAgent {
	public static void premain(String agentArgs, Instrumentation inst) {
		try {
			Class<?> agentClass = Class.forName("org.mockito.internal.PremainAttach");
			Method premainMethod = agentClass.getMethod("premain", String.class, Instrumentation.class);
			premainMethod.invoke(null, agentArgs, inst);
		} catch (Exception e) {
			throw new RuntimeException("Failed to load Mockito agent", e);
		}
	}
}
